### DISCUSSION TITLE
MQTT publish error after lightsleep on ESP32

### DISCUSSION THREAD

**Discussion by wpyoga** [micropython/micropython#11920] [ESP32] [answered]
I am experiencing issues with lightsleep. Specifically, **MQTT publish errors**. This might be related to #11877 by @davefes .

I installed [esp32-ota-20230426-v1.20.0.bin](https://micropython.org/download/esp32-ota/) onto an ESP32 dev board (ESP-WROOM-32). Haven't tested with 32D and 32E modules.

I installed [umqtt.simple](https://github.com/micropython/micropython-lib/tree/master/micropython/umqtt.simple) using `mip`.

This is a script to reproduce the issue, you can name it `lightsleep.py`:

```
import network
import machine
import umqtt.simple
# import time

sta_if = network.WLAN(network.STA_IF)

print("before disconnect from wifi")
try:
    sta_if.disconnect()
except:
    pass

print("before deactivate wifi")
sta_if.active(False)
print("before activate wifi")
sta_if.active(True)
print("before connect to wifi")
sta_if.connect("MyHomeSSID", "WifiPassword")

print("wait until connected to wifi")
while not sta_if.isconnected():
    machine.idle()

m = umqtt.simple.MQTTClient("client_hello_world", "192.168.1.111", keepalive=5)

print("connect to mqtt server")
m.connect()
while True:
    print("before publish")
    m.publish("/hello", "world", retain=False, qos=1)
    print("before lightsleep")
    machine.lightsleep(1000)
    # time.sleep(1)
    print("after lightsleep")
```

The issue can be seen like this:

```
$ mpremote soft-reset repl --inject-file lightsleep.py 
Connected to MicroPython at /dev/ttyUSB0
Use Ctrl-] or Ctrl-x to exit this shell
Use Ctrl-K to inject file "lightsleep.py"
>
MicroPython v1.20.0 on 2023-04-26; 4MB/OTA module with ESP32
Type "help()" for more information.
>>> Injecting lightsleep.py
before disconnect from wifi
before deactivate wifi
before activate wifi
before connect to wifi
wait until connected to wifi
connect to mqtt server
before publish
before lightsleep
after lightsleep
before publish
[04]Traceback (most recent call last):
  File "<stdin>", line 31, in <module>
  File "umqtt/simple.py", line 144, in publish
  File "umqtt/simple.py", line 179, in wait_msg
OSError: [Errno 113] ECONNABORTED
[04]>
MicroPython v1.20.0 on 2023-04-26; 4MB/OTA module with ESP32
Type "help()" for more information.
>>> 
```

Notes:
- `192.168.1.111` is my laptop's IP address. It has an [eclipse-mosquitto](https://github.com/eclipse/mosquitto) MQTT server running inside a docker container.
- The published message is received properly by the server.
- This happens after a soft reset, and also on a fresh boot.
- I checked the free memory size using `gc.free_mem()`, it reports around 100KB of free memory.

The error does not happen if:
- `qos=0` is used instead.
- `time.sleep(1)` is used instead of `machine.lightsleep(1000)`.

Any ideas? I'm not really familiar with Micropython, but I'm pretty comfortable with Python. So I will be able to test out any code/patch on my hardware.

Thanks!

**Comment by davefes** [comment]
For me doing a `print(f'lora is: {lora}')` gave the object location.  Today I discovered `id()`
Maybe, see if something (object location?) changes with your `m` variable, before and after `lightsleep()`.

**Reply by peterhinch** [reply]
Another possibility is if lightsleep causes the socket to be closed.

**Comment by karfas** [comment, marked as answer]
The  [ESP-IDF docs](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/system/sleep_modes.html), only one google search away, clearly state that "In Deep-sleep and Light-sleep modes, the wireless peripherals are powered down. Before entering Deep-sleep or Light-sleep modes, the application must disable Wi-Fi and Bluetooth ... ". 

I don't think it's possible to maintain a WIFI connection during any of the ESP32 sleep modes, but you can try the WIFI power save options available with recent nightly micropython builds.

**Reply by glenn20** [reply]
Yes. This was also discussed some more in #9004. You have to deinit the wifi before deep or light sleep.

Note, this has significant power implications as the esp32 uses a significant amount of power to reconnect to the AP after sleeping - potentially more than you save from the light sleep. You can improve that significantly by using a static IP (eg. see https://github.com/glenn20/upy-esp32-experiments/tree/main/ESPNowvsWifiEnergyUsage) instead of dhcp and PR #11819 to set the starting channel when reconnecting.

**Reply by wpyoga** [reply]
Ah, my Google skills are lacking. I was only searching for lightsleep() causing issues with Micropython. Thanks for the pointer.
